flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh

```diff
@@ -20,72 +20,136 @@
+    local previous_number_of_values=-1
+
+    while [[ ${number_of_values} -lt ${expected_number_of_values} ]]; do
+        if [[ ${seconds_elapsed} -ge ${polling_timeout} ]]; then
+            echo "Did not produce expected number of values within ${polling_timeout}s"
+            exit 1
+        fi
+
+        truncate_files_with_valid_data
+
+        sleep ${polling_interval}
+        ((seconds_elapsed += ${polling_interval}))
+
+        number_of_values=$(get_total_number_of_valid_lines)
+        if [[ ${previous_number_of_values} -ne ${number_of_values} ]]; then
+            echo "Number of produced values ${number_of_values}/${expected_number_of_values}"
+            previous_number_of_values=${number_of_values}
+        fi
+    done
+}
+
+function truncate_files_with_valid_data() {
+  # get truncate information
+  # e.g. "xxx xxx DEBUG xxx.BucketingSink  - Writing valid-length file for xxx/out/result8/part-0-0 to specify valid length 74994"
+  LOG_LINES=$(grep -rnw $LOG_DIR -e 'Writing valid-length file')
+
+  # perform truncate on every line
+  echo "Truncating buckets"
+
+  while read -r LOG_LINE; do
+    PART=$(echo "$LOG_LINE" | awk '{ print $10 }' FS=" ")
+    LENGTH=$(echo "$LOG_LINE" | awk '{ print $15 }' FS=" ")
+
+    if [[ -z "${PART}" ]]; then
+        continue
+    fi
+    re='^[0-9]+$'
+    if ! [[ ${LENGTH}  =~ $re ]]; then
+        continue
+    fi
+
+    echo "TRUNCATING $PART"
+
+    dd if=$PART of="$PART.truncated" bs=$LENGTH count=1 >& /dev/null
+    rm $PART
+    mv "$PART.truncated" $PART
+  done <<< "$LOG_LINES"
+}
+
+function bucketing_cleanup() {
+  stop_cluster
+  $FLINK_DIR/bin/taskmanager.sh stop-all
+}
+
+# Fix the necessary configuration parameters.
+
+set_conf_ssl
+set_config_key "heartbeat.timeout" "70000"
 
 # enable DEBUG logging level for the BucketingSink to retrieve truncate length later
 echo "" >> $FLINK_DIR/conf/log4j.properties
 echo "log4j.logger.org.apache.flink.streaming.connectors.fs.bucketing.BucketingSink=DEBUG" >> $FLINK_DIR/conf/log4j.properties
 
-set_conf_ssl
+# Start the experiment.
+
 start_cluster
 $FLINK_DIR/bin/taskmanager.sh start
 $FLINK_DIR/bin/taskmanager.sh start
 $FLINK_DIR/bin/taskmanager.sh start
 
-function bucketing_cleanup() {
+wait_for_number_of_running_tms 4
 
-  stop_cluster
-  $FLINK_DIR/bin/taskmanager.sh stop-all
-}
 on_exit bucketing_cleanup
 
-JOB_ID=$($FLINK_DIR/bin/flink run -d -p 4 $TEST_PROGRAM_JAR -outputPath $TEST_DATA_DIR/out/result \
+JOB_ID=$($FLINK_DIR/bin/flink run -d -p 4 $TEST_PROGRAM_JAR -outputPath $JOB_OUTPUT_DIR \
   | grep "Job has been submitted with JobID" | sed 's/.* //g')
 
 wait_job_running ${JOB_ID}
 
 wait_num_checkpoints "${JOB_ID}" 5
 
-echo "Killing TM"
-
-# kill task manager
+echo "Killing 1 TM"
 kill_random_taskmanager
+wait_for_number_of_running_tms 3
 
-echo "Starting TM"
-
-# start task manager again
+echo "Restarting 1 TM"
 $FLINK_DIR/bin/taskmanager.sh start
+wait_for_number_of_running_tms 4
 
-echo "Killing 2 TMs"
+sleep 10
```

Comment:
why is this additional sleep required?

Reply:
It's to allow the job to make some progress after killing the TaskManager(s)